### PR TITLE
Remove macro in iter_to_array for List

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Ref #7988

### CODE REVIEW

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -5067,76 +4680,45 @@ mod tests {
         assert_eq!(array, &expected);
     }
 
-    #[test]
-    fn test_nested_lists() {
-        // Define inner list scalars
-        let a1 = ListArray::from_iter_primitive::<Int32Type, _, _>(vec![Some(vec![
```
Comment (jayzhan211) [inline datafusion/common/src/scalar.rs:5073]: This test is length > 1, which violate the ScalarValue::List constraint. So I rewrite the test.

File: datafusion/common/src/scalar.rs [hunk 2]
```diff
@@ -1368,103 +1366,26 @@ impl ScalarValue {
...
+            let arrays_data = arrays.iter().map(|arr| arr.to_data()).collect::<Vec<_>>();
+
+            let arrays_ref = arrays_data.iter().collect::<Vec<_>>();
+            let mut mutable =
+                MutableArrayData::with_capacities(arrays_ref, false, capacity);
+
+            // ScalarValue::List contains a single element ListArray.
+            for i in 0..arrays.len() {
+                mutable.extend(i, 0, 1);
+            }
```
Comment (viirya) [inline datafusion/common/src/scalar.rs:1386]: Doesn't it need to handle null value like the macro did?

File: datafusion/common/src/scalar.rs [hunk 3]
```diff
@@ -3616,6 +3216,19 @@ mod tests {
             Some(vec![Some(4), Some(5)]),
         ]);
         assert_eq!(list_array, &expected);
+
+        let scalars = build_list::<i64>(vec![
+            vec![Some(1), Some(2), Some(3)],
+            vec![Some(4), Some(5)],
+        ]);
```
Comment (viirya) [inline datafusion/common/src/scalar.rs:3223]: Maybe we can add tests that contain null?

Comment (jayzhan211) [reply datafusion/common/src/scalar.rs:3223]: I dont think is possible to create list with null? Maybe List(NullArray)? But it is different than List(I64::None). So I did not add it.

File: datafusion/common/src/scalar.rs [hunk 4]
```diff
@@ -1368,103 +1366,36 @@ impl ScalarValue {
...
-                        }
-                        sv => {
-                            return _internal_err!(
-                                "Inconsistent types in ScalarValue::iter_to_array. \
-                                    Expected List, got {:?}",
-                                sv
-                            )
-                        }
-                    }
+        fn build_list_array(
```
Comment (alamb) [inline datafusion/common/src/scalar.rs:1369]: 😍 Amazing @jayzhan211 

I believe as written, this function will always return `ListArray` (never `LargeListArray`) so I think this probably needs to be templated on `<O: OffsetSizeTrait>` -- the way @Weijun-H  has done in https://github.com/apache/arrow-datafusion/pull/8322 for example)

File: datafusion/common/src/scalar.rs [hunk 5]
```diff
@@ -1368,103 +1366,36 @@ impl ScalarValue {
...
+            scalars: impl IntoIterator<Item = ScalarValue>,
+        ) -> Result<ArrayRef> {
+            let arrays = scalars
```
Comment (alamb) [inline datafusion/common/src/scalar.rs:1372]: Is it necessary to make a 1 element array for each ScalarValue?

I wonder if you could make the values array directly like

```rust
let values = ScalarValue::iter_to_array(scalars);
// and then make a single offset of 0
```

🤔

**Review by alamb — COMMENTED** [review]
Thank you @jayzhan211  -- this is looking like a great start

File: datafusion/common/src/scalar.rs (hunk 4 above)
Comment (jayzhan211) [reply datafusion/common/src/scalar.rs:1369]: I think mutable handle large list too, otherwise the large list test should fail 🤔

File: datafusion/common/src/scalar.rs (hunk 5 above)
Comment (jayzhan211) [reply datafusion/common/src/scalar.rs:1372]: After iter_to_array, the values is not guaranteed to be single element, so I don't think it works

Comment (jayzhan211) [reply datafusion/common/src/scalar.rs:1372]: Since the result of iter_to_array is not guaranteed to be single array like what ScalarValue is, so recursive calls in this function does not make sense to me.

File: datafusion/common/src/scalar.rs (hunk 4 above)
Comment (alamb) [reply datafusion/common/src/scalar.rs:1369]: You are right

File: datafusion/common/src/scalar.rs (hunk 5 above)
Comment (alamb) [reply datafusion/common/src/scalar.rs:1372]: I see what you are saying -- sorry I was confused.

**Comment by alamb** [comment]
I also did some ad hoc performance testing and this change did not change the performance for count distinct queries for me:

```
andrewlamb@Andrews-MacBook-Pro:~/Downloads$ /Users/andrewlamb/Software/arrow-datafusion/datafusion-cli/target/release/datafusion-cli -c 'select count(distinct trace_id), count(distinct "span.kind") from traces group by "service.name";'
DataFusion CLI v33.0.0
+---------------------------------+----------------------------------+
| COUNT(DISTINCT traces.trace_id) | COUNT(DISTINCT traces.span.kind) |
+---------------------------------+----------------------------------+
| 30714                           | 1                                |
| 30678                           | 1                                |
| 30714                           | 1                                |
| 30678                           | 1                                |
| 30673                           | 1                                |
| 3632710                         | 2                                |
+---------------------------------+----------------------------------+
6 rows in set. Query took 1.771 seconds.
```

And I verified (by putting a println in ) that this build array was getting called
